fix `require.resolve` not considering paths `.` and `..` relative

dario-piotrowicz opened the pull request:
Fixes: https://github.com/nodejs/node/issues/47000

<!--
Before submitting a pull request, please read:

- the CONTRIBUTING guide at https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md
- the commit message formatting guidelines at
  https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

If you believe this PR should be highlighted in the Node.js CHANGELOG
please add the `notable-change` label.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

dario-piotrowicz commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -1978,6 +1963,19 @@ function createRequire(filename) {
+ * @returns {boolean} true if the path is relative, false otherwise
+ */
+function isRelative(path) {
+  return path === '.' || path === '..' ||
+  StringPrototypeStartsWith(path, './') ||
+  StringPrototypeStartsWith(path, '../') ||
+  ((isWindows && StringPrototypeStartsWith(path, '.\\')) ||
+  StringPrototypeStartsWith(path, '..\\'));
```
I realize that charcode comparisons might be more performant, but I believe that this is leagues better in terms of readability, I'm happy to go with charcode comparisons if perf is a concern here

aduh95 replied:
You could probably save some cycle by returning early if `path.length && path[0] === '.'` is false

dario-piotrowicz replied:
oh yeah of course! 🙂 

Added! thanks for the suggestion 🙂

ljharb commented:
do all of these added tests fail on main, or only a subset of them?

aduh95 commented:
The `src:` subsystem is reserved for changes in `src/`. Let's use `module:` instead

dario-piotrowicz commented:
> do all of these added tests fail on main, or only a subset of them?

only a subset, but I think that the `paths` option is actually never tested anywhere else (or at least I think?) so I figured that adding some would be beneficial (without going overboard)

![Screenshot 2025-01-24 at 00 07 00](https://github.com/user-attachments/assets/535f26d1-90f6-41a9-9bdc-ff0ff70fd80b)

ljharb commented:
ah ok, so this is specifically a bug with `paths` only?

dario-piotrowicz commented:
> The `src:` subsystem is reserved for changes in `src/`. Let's use `module:` instead

sorry, my bad, updated 🫡

dario-piotrowicz commented:
> ah ok, so this is specifically a bug with `paths` only?

Yes, as far as I can tell that's the case, this for example:
https://github.com/nodejs/node/blob/869ea331f3a8215229290e2e6038956874c382a6/test/parallel/test-require-dot.js#L11-L12
does indicate that `.` is correctly treated as a relative path in standard `require` calls

aduh95 commented on test/parallel/test-require-resolve-opts-paths-relative.js:
```diff
@@ -0,0 +1,43 @@
+
+// Parent directory paths (`..`) work as intended
+{
+  assert(require.resolve('.', { paths: ['../'] }).endsWith('index.js'));
```
Can we use `assert.match` to get better output in case of error – also, let's include slash/backslash in the check while we're at it
```suggestion
  assert.match(require.resolve('.', { paths: ['../'] }), /[/\\]index.js$/);
```

dario-piotrowicz replied:
definitely! :rocket:

aduh95 commented:
May I suggest to not include `add early return for perf gain` and `fix linting` in the commit message – I would also suggest to not bother with the `Fixes:` which will be added by the bot upon landing as long as it's in the PR description.

The commit message needs to be shorter, e.g. ``module: fix relative path handling in `require.resolve` ``.

dario-piotrowicz commented:
> May I suggest to not include `add early return for perf gain` and `fix linting` in the commit message

Sorry I added them by mistake when I squashed fixup commits, I've already removed them

>  I would also suggest to not bother with the `Fixes:` which will be added by the bot upon landing as long as it's in the PR description.

Sorry the contributing MD made me thing I was supposed to add that 😓

> The commit message needs to be shorter, e.g. `` module: fix relative path handling in `require.resolve` ``.

Yes I did notice the linting error, hopefully my newer title is ok?

aduh95 commented:
> Sorry the contributing MD made me thing I was supposed to add that 😓

To be clear, you can 100% do that if you prefer to do it – what's important is for the commit landing on `main` to contain it, and what I'm saying is you can let the bot add it for you

dario-piotrowicz commented:
> > Sorry the contributing MD made me thing I was supposed to add that 😓
> 
> To be clear, you can 100% do that if you prefer to do it – what's important is for the commit landing on `main` to contain it, and what I'm saying is you can let the bot add it for you

Got it 🙂 

Maybe I can open a quick PR to add a note about that here?
https://github.com/nodejs/node/blob/869ea331f3a8215229290e2e6038956874c382a6/doc/contributing/pull-requests.md?plain=1#L183-L191

WDYT? (I'm guessing that `Fixes` is autopopulated by bots but `Refs` is still on the developer?)

dario-piotrowicz commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -1978,6 +1963,21 @@ function createRequire(filename) {
+function isRelative(path) {
+  if (StringPrototypeCharCodeAt(path, 0) !== CHAR_DOT) { return false; }
+
+  return path === '.' || path === '..' ||
```
sure, sounds good thanks 🙂

mcollina reviewed (approved):
lgtm

jsumners-nr commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -1978,6 +1963,21 @@ function createRequire(filename) {
   return createRequireFromPath(filepath);
 }
 
+/**
+ * Checks if a path is relative
+ * @param {string} path the target path
+ * @returns {boolean} true if the path is relative, false otherwise
+ */
+function isRelative(path) {
+  if (StringPrototypeCharCodeAt(path, 0) !== CHAR_DOT) { return false; }
+
+  return path.length === 1 || path === '..' ||
+  StringPrototypeStartsWith(path, './') ||
+  StringPrototypeStartsWith(path, '../') ||
+  ((isWindows && StringPrototypeStartsWith(path, '.\\')) ||
+  StringPrototypeStartsWith(path, '..\\'));
+}
```
🎉

legendecas commented on test/parallel/test-require-resolve-opts-paths-relative.js:
```diff
@@ -0,0 +1,43 @@
+const assert = require('assert');
+const fixtures = require('../common/fixtures');
+
+if (!common.isMainThread)
```
Did this PR pass the CI? This is causing CI failures on the main branch: https://github.com/nodejs/node/actions/runs/12969358219/job/36173299502

legendecas replied:
https://github.com/nodejs/node/pull/56768 should fix this

dario-piotrowicz replied:
> Did this PR pass the CI? This is causing CI failures on the main branch: https://github.com/nodejs/node/actions/runs/12969358219/job/36173299502

Yes, I think all CI checks were green before the merge
